File: Coding_Problems/11_secondHighestNum.js
function findSecondHighestNumber(arr) {
  let first = -Infinity;
  let second = -Infinity;

  for (let n of arr) {
    if (n > first) {
      second = first;
      first = n;
    } else if (n > second && n !== first) {
      second = n;
    }
  }
  return second === -Infinity ? null : second;
}

console.log(findSecondHighestNumber([3, 1, 4, 1, 5, 9, 2, 6])); // 6
console.log(findSecondHighestNumber([5, 5, 5])); // null
console.log(findSecondHighestNumber([1, 2]));

File: Coding_Problems/CPP/11_secondHighestNum.cpp
#include <iostream>
#include <vector>
#include <climits>

using namespace std;

int findSecondHighestNumber(vector<int> &arr)
{
    int first = INT_MIN;
    int second = INT_MIN;

    for (int n : arr)
    {
        if (n > first)
        {
            second = first;
            first = n;
        }
        else if (n > second && n != first)
            second = n;
    }

    return second == INT_MIN ? -1 : second;
}

int main()
{
    vector<int> arr = {3, 1, 4, 1, 5, 6, 2, 6};

    cout << findSecondHighestNumber(arr);

    return 0;
}
